### ISSUE TITLE
Redux Integration (ControlledRouter)

### ISSUE THREAD

**Issue by donnanicolas** [ReactTraining/react-router#3879]
Hi, we are starting to migrate from Angular to React+Redux, and we are planning on using a routing system, but the old version of ReactRouter had some issues when using Redux. V4 seems a lot better for that, and in the faqs you talk about a ControlledRouter close to being published, but I can seem to find it.

We (me and the other developers) would like to help you with the development, since is essential to out product, where can we find the *ControllerRouter* and how can we help??

Cheers

PS: I mention @JulianMayorga since he is one of my teammates.

**Comment by timdorr** [comment]
We don't have anything here. But it would be a great addon library. I would be more than happy to give up the react-router-redux repo or have it live on that project.

**Comment by nhagen** [comment]
From the FAQ:

> We have a <ControlledRouter> close to being published that makes redux integration with React Router the same as ... uh ... integrating an <input> with Redux.

I'm a little confused now.

**Comment by timdorr** [comment]
@nhagen There's nothing here, as in "on this repo" here. Addons like ControlledRouter or static, top-level routes are not going into the core of 4.0. They will be externalized in same way. This might end up being a monorepo (Babel-style) or they might get their own separate repos. Nothing's been decided yet and all of this stuff is still very much in progress.

**Comment by mjackson** [comment]
I know that @ryanflorence has done some work on a `<ControlledRouter>` and was planning to release it as an addon. But we could possibly put it in the `react-router-redux` project too.

> I'm a little confused now.

Don't be :) The API for a `<ControlledRouter>` looks like this:

```js
<ControlledRouter location={ pathname: '/' }/>
```

You can think of `<ControlledRouter location>` like `<input value>`. When you set it, the router automatically reflects that value.

**Comment by donnanicolas** [comment]
We where thinking on how to implement the API that @mjackson mentioned:

```
class ControllerRouter extends Component {
  componentDidUpdate() {
    //Make some check to avoid pushing the same route many times
    browserHistory.push(this.props.location.pathname)
  }
}
```

Is it as easy as this?? 

`browserHistory#push` was the only way we found to change routes from the code

**Comment by mjackson** [comment]
If you do a `PUSH` every time you'll increase the history stack. It would be like every time you give your `<ControlledRouter>` a new `location`, someone clicked on a link, which isn't what you always want. Sometimes you may want to go back, for example. Or undo the most recent location change.

**Comment by steida** [comment]
https://gist.github.com/steida/16da36887e6e6b466e12ba3a4ae6ce87

**Comment by ryanflorence** [comment]
I have the code, just need to package it up. on vacation with family but
maybe after los niños are in bed I can do it, if not, I'll just post a gist
you can use

**Comment by steida** [comment]
Just gist us what you have and enjoy the vacation with your family unwired. There is no show stopper in RR4.

**Comment by donnanicolas** [comment]
Based on @steida's version we've created (with @JulianMayorga) a version that change react to state changes

https://gist.github.com/donnanicolas/3d76397a92551f449637590bf0413133

**Comment by timdorr** [comment]
I'll put something on react-router-redux as a 5.0 branch soon. We'll settle on a particular "best" implementation soon, but @donnanicolas has the right idea.

**Comment by ryanflorence** [comment]
sorry everybody, I just keep forgetting to put code up, at least you'll
appreciate it more after trying yourself!

**Comment by dlebedynskyi** [comment]
don't know if this is a right place to ask or should open another issue. 
tl;dr;  How to integrate async data loading on server for universal app with react-router^4 ?

common task that we now have to do is to fetch some data on server before render, put it into redux state (so using ControlledRouter it seems with ^4), then sync stuff on client. 

as one lib that does something like this - Redial. https://github.com/markdalgleish/redial 
This is not the only option out there but basically all of them depend on match and renderComponents. 

Would be great if you guys supply some way of supporting this. Maybe someway to promisify render for ControlledRouter?

**Comment by steida** [comment]
@dlebedynskyi I do this https://github.com/este/este/commit/47efea50fcee5761cf406f2240bdeda8f75c1031 No library is needed.

**Comment by timdorr** [comment]
Tracking a possible implementation here: reactjs/react-router-redux#460 Thoughts, improvements, major refactors are all welcome!

**Comment by ngxiaoyi** [comment]
Hi @dlebedynskyi ,

Just shared my codes which build universal app successfully with react-router^4 & redux:
`// server.js
...
import Express from 'express'
import { ServerRouter, createServerRenderContext } from 'react-router'
...
const app = Express()
app.get('/team', handleRender)

function handleRender(req, res){
Promise.then(function(response){
  let initialState = {...}
  const context = createServerRenderContext()
  let store = configureStore(initialState)
  let html = renderToString(
      <Provider store={store}>
        <ServerRouter
          location={req.url}
          context={context}
        >
          <Team />
        </ServerRouter>
      </Provider>
    )
    res.send(renderPage(html, initialState))
})
}
`
`// client.js
...
import { BrowserRouter } from 'react-router'
...
const initialState = window.__INITIAL_STATE__
let store = configureStore(initialState)
let rootElement = document.getElementById('root')
ReactDOM.render(
  <Provider store={store}>
      <BrowserRouter>
        <Team />
      </BrowserRouter>
  </Provider>,
  rootElement
)
`

`// container/Team.js
export default class Team extends Component {
  constructor(props) {
    super(props)
  }

  render() {
    return (
      <div>
        <Match pattern="/team" exactly component={TeamList} />
        <Match pattern="/team/:teamId" component={TeamItem} />
      </div>
    )
  }
}
`

It works as expected if i visit one of the link under TeamList component which points to '/team/123'

**Comment by dlebedynskyi** [comment]
@steida Thanks this looks like a good approach. I'll check it out

**Comment by gaelduplessix** [comment]
Created a `ControlledRouter` based on @steida's gist: https://gist.github.com/gaelduplessix/050e5cce31f9fabb1030f4ba47663db5

Currently porting my React/Redux app with it, seems to work fine :)

**Comment by kevinold** [comment]
FYI, Ryan posted his `ControlledRouter` here: https://github.com/ReactTraining/react-router/issues/3905#issuecomment-250991161

**Comment by alisd23** [comment]
@dlebedynskyi I made a library for exactly what you're looking for (I think). I had the same issue but with MobX - should work for either. https://github.com/alisd23/react-async-actions

**Comment by bdudelsack** [comment]
Can anybody please provide an example how to integrate the ControlledRouter from Ryan (#3905)  with Redux?

**Comment by steida** [comment]
@bdudelsack I don't believe that piece of code really works also RR4alpha4 has its own issue. I recommend do not migrate yet.

**Comment by dlebedynskyi** [comment]
@alisd23 yeah we ended up with something similar. note that react is strongly thinking of ditching componentWillMount in favor of constructor.

**Comment by ryanflorence** [comment]
A real controlled router is coming, the enthusiasm is great but please
remember we're still in alpha with this stuff.

**Comment by gaelduplessix** [comment]
Yup, thanks for your work @ryanflorence but I think we've been a bit too enthusiast right now. We already migrated some of our code to v4  (and loved it!) but I guess we're gonna have to revert to the previous version for now :(

**Comment by timdorr** [comment]
This is now [available as an addon](https://github.com/ReactTraining/react-router-addons-controlled). I've also got [the next version of react-router-redux](https://github.com/reactjs/react-router-redux/pull/460) started on a similar path. So, we've got some options for everyone now.
